File: Coursework/Data Structures and Algorithms/Implementations/Algorithms/Count Sort.java
// Counting Sort, O(n + mx), stable, out of place

public class Main {
    public static void main(String[] args) {

        int[] x = {2, 5, 3, 0, 2, 3, 0, 3};               // Array to be sorted, no negative/huge numbers.
        int mx = -1;                                      // To hold the max value in array.

        for (int i : x) mx = Math.max(mx, i);             // Find max value.
        
        int[] c = new int[mx + 1], y = new int[x.length]; // Init counting array, result array.
        
        for (int i : x) c[i]++;                           // Fill count array.
        for (int i = 1; i <= mx; i++) c[i] += c[i - 1];   // Do a prefix sum on counting array.
        
        for (int i : x) {                                 
            y[c[i] - 1] = i;                              // Put each array element in the correct index.
            c[i]--;                                       // Decrement the counter.
        }
        
        for (int i : y) System.out.print(i + " ");        // Output the result.
    }
}

File: Bachelors/Data Structures and Algorithms/Implementations/Algorithms/Count Sort.cpp
#include <bits/stdc++.h>
using namespace std;

/// Count Sort O(n + mx), stable, out of place
/// Won't work for negative numbers or numbers with high max.

int main()
{
    int x[] = {5,4,5,2,1};
    int n = sizeof(x)/sizeof(x[0]);

    int mx = *max_element(x,x+n);           /// O(n)

    int m[mx+1], y[n];
    memset(m,0,sizeof(m));                  /// O(n)

    for(int i=0;i<n;i++) m[x[i]]++;         /// O(n)

    for(int i=1;i<=mx;i++) m[i]+=m[i-1];    /// O(mx)

    for(int i=0;i<n;i++){                   /// O(n)
        y[m[x[i]]-1]=x[i];
        m[x[i]]--;
    }

    for(int u:y)cout<<u<<" ";
}